solutionbox/inception/datalab_solutions/inception/_cloud.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Cloud implementation for preprocessing, training and prediction for inception model.
+"""
+
+import apache_beam as beam
+import base64
+import collections
+import datetime
+from googleapiclient import discovery
+import google.cloud.ml as ml
+import logging
+import os
+
+from . import _model
+from . import _preprocess
+from . import _trainer
+from . import _util
+
+
+_CLOUDML_DISCOVERY_URL = 'https://storage.googleapis.com/cloud-ml/discovery/' \
+                         'ml_v1beta1_discovery.json'
+_TF_GS_URL= 'gs://cloud-datalab/deploy/tf/tensorflow-0.12.0rc0-cp27-none-linux_x86_64.whl'
+
+
+class Cloud(object):
+  """Class for cloud training, preprocessing and prediction."""
+
+  def __init__(self, project, checkpoint=None):
+    self._project = project
+    self._checkpoint = checkpoint
+    if self._checkpoint is None:
+      self._checkpoint = _util._DEFAULT_CHECKPOINT_GSURL
+
+  def preprocess(self, input_csvs, labels_file, output_dir, pipeline_option=None):
+    """Cloud preprocessing with Cloud DataFlow."""
+
+    job_name = 'preprocess-inception-' + datetime.datetime.now().strftime('%y%m%d-%H%M%S')
+    options = {
+        'staging_location': os.path.join(output_dir, 'tmp', 'staging'),
+        'temp_location': os.path.join(output_dir, 'tmp'),
+        'job_name': job_name,
+        'project': self._project,
+        'extra_packages': [ml.sdk_location, _util._PACKAGE_GS_URL, _TF_GS_URL],
+        'teardown_policy': 'TEARDOWN_ALWAYS',
+        'no_save_main_session': True
+    }
+    if pipeline_option is not None:
+      options.update(pipeline_option)
+
+    opts = beam.pipeline.PipelineOptions(flags=[], **options)
+    p = beam.Pipeline('DataflowPipelineRunner', options=opts)
+    _preprocess.configure_pipeline(
+        p, self._checkpoint, input_csvs, labels_file, output_dir, job_name)
+    p.run()
+
+  def train(self, labels_file, input_dir, batch_size, max_steps, output_path, credentials,
```

Comment:
%ml train (in datalab code as a separate change) basically runs the function through inspection. The function is not supposed to be called directly.

There are a few reasons that a function is more flexible than program args:

1. Function arguments can easily be enumerated programmatically (so --dumpargs is easy to implement)
2. It can display package specific content in Datalab directly. For example, predict() can display image in output along with labels. preprocess() can display html links for DataFlow jobs. train() can display job logs. The output behavior can be different package by package.
3. There can be extra processing before submitting to cloud. In this case, we get the number of labels from labels file. We can also do more (such as validating package specific args).